login_sql.8
.\"Copyright (c) 2010, 2011, 2012, 2013 LEVAI Daniel
.\"All rights reserved.
.\"Redistribution and use in source and binary forms, with or without
.\"modification, are permitted provided that the following conditions are met:
.\"	* Redistributions of source code must retain the above copyright
.\"	notice, this list of conditions and the following disclaimer.
.\"	* Redistributions in binary form must reproduce the above copyright
.\"	notice, this list of conditions and the following disclaimer in the
.\"	documentation and/or other materials provided with the distribution.
.\"THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
.\"ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\"WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\"DISCLAIMED. IN NO EVENT SHALL LEVAI Daniel BE LIABLE FOR ANY
.\"DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\"(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
.\"LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
.\"ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\"(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\"SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.Dd Feb 11, 2013
.Dt LOGIN_SQL 8
.Os
.Sh NAME
.Nm login_sql
.Nd provides SQL authentication
.Sh SYNOPSIS
.Nm
.Op Fl s Ar service
.Ar user
.Op Ar class
.Sh DESCRIPTION
The
.Nm
utility (if configured) is called by
.Xr login 1 ,
.Xr su 1 ,
and others to authenticate the
.Ar user
against an SQL database. Currently PostgreSQL, MySQL and SQLite is supported.
.Pp
The
.Ar user
argument is the login name of the user being authenticated.
.Pp
The
.Ar service
argument specifies which protocol to use with the
invoking program.
.Pp
The auth style name is
.Ql sql .
You can use it after you configured a login class in
.Xr login.conf 5
to use this auth style, and created a configuration file for the
.Nm
authentication program.
.Ss LOGIN.CONF VARIABLES
The
.Nm
utility uses only the following
.Em optional
.Xr login.conf 5
variable:
.Bl -tag -offset ||| -width |
.It Cm x-sql-cfg_file
Specifies the path name for the configuration file. If you omit this variable, then the default configuration file will be used.
.El
.Ss CONFIGURATION FILE
The configuration file ignores any line which doesn't start with a recognized configuration option. Options are in the form of
.Em option=value .
.Pp
Global configuration options:
.Bl -tag -offset ||| -width |
.It Cm sql_backend
The SQL backend to use for retrieving the password. Valid values are
.Cm pgsql,
.Cm mysql
and
.Cm sqlite.
.It Cm db_host
Host name or IP address (on backends where applicable).
.It Cm db_port
Port number for a TCP/IP connection (on backends where applicable).
.It Cm db_name
The database name (or a filename with such backends) which contains the table with the username and password.
.It Cm db_username
Username used for connecting to the database (on backends where applicable).
.It Cm db_password
Password used for connecting to the database (on backends where applicable).
.It Cm db_table
The table name in the database which contains the username and password columns.
.It Cm column_username
The column's name which contains the username.
.It Cm column_password
The column's name which contains the password.
.It Cm column_scheme
The column's name which contains the password scheme. If not empty nor NULL, this scheme overrides the global
.Cm pw_scheme
option.
.It Cm column_enabled
(Default: )
.Pp
The column's name which contains the enabled variable. This can be used to control if a user is allowed to login or not. This is compared to value
.Ql true
on backends that support the boolean type, and to 1 where it is represented as an integer. If this parameter is omitted or its value is empty (default), don't use an enabled column when authenticating a user.
.It Cm pw_scheme
The message digest or encryption algorithm to use with the user supplied password. This will be the default, which can be overridden with a
.Cm column_scheme
column in the database table. If the
.Cm column_scheme
column's value is an empty string or NULL, then this will be used for the username in question.
.Pp
Valid digests are the following:
.Bl -tag -offset ||| -width |
.It Ar smd4 smd5 smdc2 sripemd160 ssha1 ssha224 ssha256 ssha384 ssha512 swhirlpool
These are various salted message digests.
.Xr login_sql 8
comes with a utility called
.Xr login_sql_pw 1
that can generate such digests.
.It Ar blowfish md5crypt
These digests use the
.Xr crypt 3
function to encrypt the user supplied password with salted
.Em blowfish
or salted
.Em md5
algorithms, respectively. The
.Xr encrypt 1
utility can be used to generate such digests.
.It Ar cleartext
This means that the password is stored in clear text in the database, so no hashing is necessary for the password entered.
.It OpenSSL digests
Furthermore you can use any digest which is supported by your OpenSSL library. For more information, enter
.Dl openssl dgst -h
on the command line.
.El
.It Cm empty_password
(Default: no)
.Pp
If
.Cm yes
then allow comparison of empty passwords. If
.Cm no
(default)
.Nm
refuses to authorize a login if either the supplied or queried password is empty.
.Em Note
that with eg. OpenSSH, there is also a separate sshd_config option to permit empty passwords.
.El
.Pp
.Em PostgreSQL
specific options:
.Bl -tag -offset ||| -width |
.It Cm pgsql_dbconnection
This is a space separated list of keyword=value pairs that can be used as extra connection parameters for PostgreSQL.
.Em Note
that host=, port=, dbname=, user= and password= PostgreSQL connection options are specified with the global configuration options (see above).
.El
.Bd -ragged -offset |||
For more information on the available extra options:
.Pp
http://www.postgresql.org/docs/9.1/static/libpq-connect.html
.Pp
Search for PQconnectdbParams, and refer to the recognized parameters.
.Ed
.Pp
.Em MySQL
specific options:
.Bl -tag -offset ||| -width |
.It Cm mysql_key
Path name to the key file.
.It Cm mysql_cert
Path name to the certificate file.
.It Cm mysql_ca
Path name to the certificate authority file.
.It Cm mysql_capath
Path name to the directory that contains trusted SSL CA certificates in pem format.
.It Cm mysql_cipher
List of permissible ciphers to use for SSL encryption.
.El
.Bd -ragged -offset |||
For the MySQL parameters' valid or default values please read the MySQL documentation here:
.Pp
http://dev.mysql.com/doc/refman/5.5/en/mysql-real-connect.html
.Pp
For the MySQL SSL parameters' valid or default values please read the MySQL SSL documentation here:
.Pp
http://dev.mysql.com/doc/refman/5.5/en/mysql-ssl-set.html
.Ed
.Sh FILES
.Bl -hang -width ||||||||||||||||||||||||||||||||| -compact
.It Pa /etc/login_sql.conf
Default configuration file
.It Pa /etc/login.conf
To setup the
.Nm
auth style
.El
.Sh EXAMPLES
.Bl -tag -width |
.It Em login.conf :
.Bd -literal
sql:\e
	:auth=sql:\e
	:x-sql-cfg_file=/etc/login_sql.conf:\e
	:tc=default:
.Ed
.Pp
The cfg_file parameter is optional. If omitted, the default will be used, which is
.Pa /etc/login_sql.conf .
.It Em Example configuration for PostgreSQL :
.Bd -literal
sql_backend=pgsql

db_host=127.0.0.1
db_port=5432
db_name=login_sql
db_username=login_sql
db_password=login_sql
db_table=userdb

column_username=username
column_password=password
column_scheme=scheme
column_enabled=enabled

pw_scheme=ssha256

# Optional, extra PostgreSQL options:
pgsql_dbconnection=connect_timeout=10 sslmode=verify-full
.Ed
.It Em Example configuration for SQLite :
.Bd -literal
sql_backend=sqlite

db_name=/etc/login_sql.sqlite
db_table=userdb

column_username=username
column_password=password
column_scheme=scheme
column_enabled=enabled

pw_scheme=ssha256
.Ed
.It Em Generating a blowfish password with
.Xr encrypt 1 :
.Bd -literal
$ encrypt s3cr3t
$2a$06$.sqd/Wevf1C6B520zdshB.1lNPGE/FkqK3/O9p0uj7lzVofPtBpJu
.Ed
.Pp
You can use this value for a password row if you use
.Ql blowfish
for
.Cm pw_scheme .
.It Em Generating passwords with
.Xr login_sql_pw 1 :
.Bd -literal
$ login_sql_pw -a sha1 -p sha1s3cr3t
saZ6GV5M5ZPKBeBxRE2eh2FFR4mLTeDk
$ login_sql_pw -a sha256 -p sha256s3cr3t
PwsMI5v86ycSdJGudPigVdE73GbeAHhnkZaSOrpRw6KnVfvU
$ login_sql_pw -a sha512
Password:
JLacuTMLsCou5OPk92Jclw/QSIDrMqg+PIAfd3XHavoCLNp1OPm9zKPrwA7J+wVO3SCWXU5MapnBQdYFOJlP7rlr6rw=
$ login_sql_pw -a ripemd160 -p test
tdHUAgTPqRrbXcv11jQ6SIlsLzcW+10V
.Ed
.Pp
You can use these values for a password row if you use
.Ql ssha1 ,
.Ql ssha256
or
.Ql ssha512
for
.Cm pw_scheme .
.It Em Generating a salted md5 password with
.Xr encrypt 1 :
.Bd -literal
$ encrypt -m s3cr3t
$1$EpRnA1xB$mpt54XE/LOyF18ynFuM1u/
.Ed
.Pp
You can use this value for a password row if you use
.Ql md5crypt
for
.Cm pw_scheme .
.It Em Generating an sha256 hashed password with
.Xr openssl 1 :
.Bd -literal
$ printf "s3cr3t" |openssl dgst -sha256
(stdin)= 4e738ca5563c06cfd0018299933d58db1dd8bf97f6973dc99bf6cdc64b5550bd
.Ed
.Pp
You can use this value for a password row if you use
.Ql sha256
for
.Cm pw_scheme .
.It Em Sample SQL table Ql userdb :
.Bd -literal
CREATE TABLE userdb (
	username text,
	password text,
	scheme text,
	enabled boolean,
	comment text  -- optional, eg.: a description of a user
);
INSERT INTO userdb VALUES ('customer_1', 'Sup3rSecret', 'cleartext', true, 'The bakery next door');
INSERT INTO userdb VALUES ('customer_2', '4e738ca5563c06cfd0018299933d58db1dd8bf97f6973dc99bf6cdc64b5550bd', '', false, 'Flower shop');
.Ed
.Pp
The first entry uses a clear text password, and the default
.Cm pw_scheme
in the example configuration file is overridden by the filled in
.Ql scheme
column.
.Pp
The second entry uses a hashed password, and the default
.Cm pw_scheme
in the example configuration file is being used because of the empty
.Ql scheme
value. Also this account is disabled, and won't be able to login.
.El
.Sh AUTHOR
.Nm
was written by
.An LEVAI Daniel
<leva@ecentrum.hu>.
.Pp
Source, information, bugs:
http://login-sql.googlecode.com
.Sh SEE ALSO
.Xr login 1 ,
.Xr su 1 ,
.Xr login.conf 5 ,
.Xr openssl 1 ,
.Xr encrypt 1 ,
.Xr psql 1 ,
.Xr mysql 1 ,
.Xr sqlite3 1


login_sql.c
/*
 * Copyright (c) 2010, 2011, 2012, 2013 LEVAI Daniel
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *	* Redistributions of source code must retain the above copyright
 *	notice, this list of conditions and the following disclaimer.
 *	* Redistributions in binary form must reproduce the above copyright
 *	notice, this list of conditions and the following disclaimer in the
 *	documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL LEVAI Daniel BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include <sys/time.h>
#include <sys/resource.h>
#include <sys/syslimits.h>
#include <sys/types.h>

#include <login_cap.h>
#include <unistd.h>

#include "common.h"
#include "sql_check.h"
#include "login_sql.h"


int
main(int argc, char *argv[])
{
	struct		rlimit rl;
	login_cap_t	*lc = NULL;

	FILE		*back = NULL;
	int		mode = 0, c, count = 0;
	char		response[MAX_PASSWORD];
	char		*class = NULL, *username = NULL, *password = NULL;
	char		*config_file = NULL;


	rl.rlim_cur = 0;
	rl.rlim_max = 0;
	(void)setrlimit(RLIMIT_CORE, &rl);

	(void)setpriority(PRIO_PROCESS, 0, 0);

	openlog("login_sql", LOG_PID, LOG_AUTH);

	while ((c = getopt(argc, argv, "v:s:d")) != -1) {
		switch (c) {
		case 'v':
			break;
		case 's':	/* service */
			if (strncmp(optarg, "login", 5) == 0) {
				mode = MODE_LOGIN;
			} else if (strncmp(optarg, "challenge", 9) == 0) {
				mode = MODE_CHALLENGE;
			} else if (strncmp(optarg, "response", 8) == 0) {
				mode = MODE_RESPONSE;
			} else {
				syslog(LOG_ERR, "%s: invalid service", optarg);

				closelog();
				exit(AUTH_FAILED);
			}
			break;
		case 'd':
			back = stdout;
			break;
		default:
			syslog(LOG_ERR, "usage error1");

			closelog();
			exit(AUTH_FAILED);
		}
	}

	switch (argc - optind) {
		case 2:
			class = argv[optind + 1];
			/*FALLTHROUGH*/
		case 1:
			username = argv[optind];
			break;
		default:
			syslog(LOG_ERR, "usage error2");

			closelog();
			exit(AUTH_FAILED);
	}

	if (back == NULL && (back = fdopen(3, "r+")) == NULL) {
		syslog(LOG_ERR, "reopening back channel: %m");

		closelog();
		exit(AUTH_FAILED);
	}

	switch (mode) {
		case MODE_LOGIN:
			password = getpass("Password:");
			break;
		case MODE_CHALLENGE:
			fprintf(back, BI_CHALLENGE "\n");

			closelog();
			exit(AUTH_OK);
			break;
		case MODE_RESPONSE:
			/* read the first string, which is the challenge (we do not use that) */
			while ( (read(3, &response[count], (size_t)1) == 1)  &&  count < MAX_PASSWORD ) {
				if (response[count] == '\0') {	/* read terminating null char in challenge */
					break;
				}
				count++;
			}

			/* read the second string, which is the response and the password */
			count = 0;
			while ( (read(3, &response[count], (size_t)1) == 1)  &&  count < MAX_PASSWORD ) {
				if (response[count] == '\0') {	/* read terminating null char in response */
					password = response;
					break;
				}
				count++;
			}
			break;
		default:
			syslog(LOG_ERR, "%d: unknown mode", mode);

			closelog();
			exit(AUTH_FAILED);
			break;
	}

	if (!password) {
		syslog(LOG_ERR, "error reading password\n");

		closelog();
		exit(AUTH_FAILED);
	}

	/* if defined in login.conf(5), get the config file's path */
	lc = login_getclass(class);
	if (!lc) {
		syslog(LOG_ERR, "unknown class: %s\n", class);

		closelog();
		exit(AUTH_FAILED);
	}
	config_file = login_getcapstr(lc, CAP_CONFIG_FILE, NULL, NULL);
	login_close(lc);

	/* check against SQL */
	if (sql_check(username, password, config_file)) {
		fprintf(back, BI_AUTH "\n");
		syslog(LOG_NOTICE, "authorize ok for %s\n", username);

		closelog();
		exit(AUTH_OK);
	} else {
		fprintf(back, BI_REJECT "\n");
		syslog(LOG_NOTICE, "authorize fail for %s\n", username);
	}

	closelog();
	exit(AUTH_FAILED);
} /* main() */
